# 해당 파일 설명
이 코드를 모두 실행하고 나면, /convert 경로에 다음의 세 파일이 생성됩니다.

1. plant1_by_loc_with_datetime
2. plant2_by_loc_with_datetime
3. plant_test_with_datetime

## 추가 설명
위의 세 파일은, 날짜, plant, loc을 기반으로 제공 받은 데이터에 관한 feature를 정리하여 저장합니다.

# 1. Google drive Mount

In [ ]:
from google.colab import auth
auth.authenticate_user()

from google.colab import drive
drive.mount('/content/gdrive')


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


# 2. Define Directory

In [ ]:
# 사용하기 전 변경해주어야 하는 파라미터
BASE_WORKING_DIR = "/content/gdrive/My Drive/weather_bigcontest/4_code/"
#BASE_WORKING_DIR = "/content/gdrive/My Drive/Colab Notebooks/weather_bigcontest/4_code/"
USERNAME = "park"

In [ ]:
import os
# 사용하는 디렉토리 정의
DATA_BASE_PATH = os.path.join(BASE_WORKING_DIR, "data")
DATA_RAW_PATH = os.path.join(DATA_BASE_PATH, "raw")
DATA_EXPORT_PATH = os.path.join(DATA_BASE_PATH, "export")
DATA_CONV_USER_PATH = os.path.join(DATA_BASE_PATH, USERNAME)
DATA_CONV_PATH = os.path.join(DATA_BASE_PATH, "converted")
MODEL_PATH = os.path.join(BASE_WORKING_DIR, "model")
RESULT_PATH = os.path.join(BASE_WORKING_DIR, "result")

WORKING_DIR = os.path.join(BASE_WORKING_DIR, USERNAME)

print('Current Work Directory :',  os.getcwd(), '\n')
print('Username :',  USERNAME, '\n')

print("DATA_BASE_PATH : ", DATA_BASE_PATH)
print("DATA_RAW_PATH : ", DATA_RAW_PATH)
print("DATA_EXPORT_PATH : ", DATA_EXPORT_PATH)
print("DATA_CONV_USER_PATH : ", DATA_CONV_USER_PATH)
print("DATA_CONV_PATH : ", DATA_CONV_PATH)

print("MODEL_PATH : ", MODEL_PATH)

print("RESULT_PATH : ", RESULT_PATH)

print("WORKING_DIR : ", WORKING_DIR)


Current Work Directory : /content 

Username : park 

DATA_BASE_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/data
DATA_RAW_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/data/raw
DATA_EXPORT_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/data/export
DATA_CONV_USER_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/data/park
DATA_CONV_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/data/converted
MODEL_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/model
RESULT_PATH :  /content/gdrive/My Drive/weather_bigcontest/4_code/result
WORKING_DIR :  /content/gdrive/My Drive/weather_bigcontest/4_code/park


In [ ]:
# 데이터 무엇이 있는지 확인하자.
# 아래 코드를 수행하게 되면, "../input/"에서 리눅스 명령어 "ls"를 수행한 결과를 보여주게 된다. 이 때 나온 binary code를 "utf8"로 디코딩해서 보여준다.
from subprocess import check_output
print("===== Raw Data List =====\n", check_output(["ls", DATA_RAW_PATH]).decode("utf8"))
print("===== Export Data List =====\n" ,check_output(["ls", DATA_EXPORT_PATH]).decode("utf8"))
print("===== Converted User Data List =====\n" ,check_output(["ls", DATA_CONV_USER_PATH]).decode("utf8"))
print("===== Converted Data List =====\n" ,check_output(["ls", DATA_CONV_PATH]).decode("utf8"))
print("===== Model List =====\n" ,check_output(["ls", MODEL_PATH]).decode("utf8"))
print("===== Result file List =====\n" ,check_output(["ls", RESULT_PATH]).decode("utf8"))

===== Raw Data List =====
 plant1_train.csv
plant2_train.csv
plant2_train.gsheet
plant_test.csv
submit.csv
submit.gsheet

===== Export Data List =====
 dangjin_aws
seosan_asos
sinpyeong_aws
songak_forecast
songak_local
songsan_forecast
songsan_local

===== Converted User Data List =====
 backup_0.96model
data_all.csv
data_all_out.csv
data_all_out_in.csv
date_trend.csv
date_trend_point.csv
ExtraTreesClassifier_th_10.csv
GradientBoostingClassifier_all_feature_th_40.csv
lastplus_th_10.csv
last_th_08.csv
last_th_10.csv
model
model_cond_all.pkl
model_cond_all_plus_.pkl
model_hum_in.pkl
model_hum_out_loc.pkl
model_tem_coil.pkl
model_tem_in.pkl
model_tem_out_loc.pkl
plant1_by_loc.csv
plant1_by_loc.gsheet
plant2_by_loc.csv
plant_test_all_weather_feature_add_cluster_add_out.csv
plant_test_all_weather_feature_add_cluster_add_out_in.csv
plant_test_all_weather_feature_add_cluster_add_out_in.gsheet
plant_train_all_weather_feature_add_cluster.csv
predict.csv
RandomForestClassifier_th_10.csv
result.c

# 4. Import Library

In [ ]:
# 기본 데이터 라이브러리 로드
import sys # 시스템 파라미터에 접근할 수 있게 도와준다.
print("Python version: {}". format(sys.version))

import pandas as pd # 데이터 정제에 도움을 주는 라이브러리
print("pandas version: {}". format(pd.__version__))

import matplotlib # 매트랩에서 사용하는 시각화 도구를 사용할 수 있게 도와주는 시각화 도구
print("matplotlib version: {}". format(matplotlib.__version__))

import numpy as np # 행렬 계산을 위해 필요한 라이브러리
print("NumPy version: {}". format(np.__version__))

import scipy as sp # 수학 관련 함수가 내장된 라이브러리
print("SciPy version: {}". format(sp.__version__)) 

import IPython
from IPython import display # 주피터 노트북에서 예쁘게 시각화 해주는 도구
print("IPython version: {}". format(IPython.__version__)) 

import sklearn # 각종 통계 도구와 머신 러닝 알고리즘이 내장되어 있는 라이브러리
print("scikit-learn version: {}". format(sklearn.__version__))

# 파이썬 내장 라이브러리
import random
import datetime
import time
import os
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta
import copy


# Jupyter Notebook 이나 ipython 을 사용하다보면 향후 버전이 올라갈 때 변경될 사항 등을 알려주는 경고 메시지(warning message)를 뜨지 않게 해준다.
import warnings
warnings.filterwarnings('ignore')
print('-'*25)

Python version: 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0]
pandas version: 1.0.5
matplotlib version: 3.2.2
NumPy version: 1.18.5
SciPy version: 1.4.1
IPython version: 5.5.0
scikit-learn version: 0.22.2.post1
-------------------------


In [ ]:
# 모델링 라이브러리, 수학 계산 툴 로드
# 일반적인 모델링 라이브러리
from sklearn import svm, tree, linear_model, neighbors, naive_bayes, ensemble, discriminant_analysis, gaussian_process
from xgboost import XGBClassifier

# 모델링 시 헬퍼 함수들
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn import feature_selection
from sklearn import model_selection
from sklearn import metrics

# interpolation 함수
import scipy.interpolate as spi

# model load
import pickle
from sklearn.externals import joblib

# 시각화 도구
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.pylab as pylab
import seaborn as sns
from pandas.plotting import scatter_matrix

# 시각화 도구 default 세팅
# 주피터 노트북에서 plot 결과를 볼 수 있게 해준다.
%matplotlib inline
mpl.style.use('ggplot') # matplotlib에서 plot되는 결과를 선택할 수 있다.
sns.set_style('white') # seaborn에서 사용할 style을 설정할 수 있다.
pylab.rcParams['figure.figsize'] = 12,8 # plot의 크기와 선 등의 기본 값을 설정할 수 있다.

# 시각화 Style 설정하기
# print(plt.style.available)
# ['bmh', 'classic', 'dark_background', 'fast', 'fivethirtyeight', 'ggplot', 'grayscale', 'seaborn-bright', 'seaborn-colorblind', 'seaborn-dark-palette', 'seaborn-dark', 'seaborn-darkgrid', 'seaborn-deep', 'seaborn-muted', 'seaborn-notebook', 'seaborn-paper', 'seaborn-pastel', 'seaborn-poster', 'seaborn-talk', 'seaborn-ticks', 'seaborn-white', 'seaborn-whitegrid', 'seaborn', 'Solarize_Light2', 'tableau-colorblind10', '_classic_test']

plt.style.use(['fivethirtyeight'])
sns.set_style('darkgrid')

# 5. Read Raw Data

In [ ]:
# 데이터 나누기
# 우리는 데이터를 3단계로 나눠야 한다.
# 1. train data
# 2. test data(만든 모델을 test)
# 3. validation(최종 예측 용)
data_raw_plant1 = pd.read_csv(os.path.join(DATA_RAW_PATH, "plant1_train.csv"))
data_raw_plant2 = pd.read_csv(os.path.join(DATA_RAW_PATH, "plant2_train.csv"))
data_val  = pd.read_csv(os.path.join(DATA_RAW_PATH, "plant_test.csv"))

# unnamed: 0 제거
data_raw_plant1.drop(['Unnamed: 0'], axis='columns', inplace=True)
data_raw_plant2.drop(['Unnamed: 0'], axis='columns', inplace=True)
data_val.drop(['Unnamed: 0'], axis='columns', inplace=True)

# column 명 변경
data_raw_plant1.columns = ['mea_ddhr', 'tem_in_loc1','hum_in_loc1', 'tem_coil_loc1','tem_in_loc2', 'hum_in_loc2','tem_coil_loc2', 'tem_in_loc3','hum_in_loc3', 'tem_coil_loc3','tem_out_loc1', 'hum_out_loc1','cond_loc1', 'cond_loc2','cond_loc3']
data_raw_plant2.columns = ['mea_ddhr', 'tem_in_loc1','hum_in_loc1', 'tem_coil_loc1','tem_in_loc2', 'hum_in_loc2','tem_coil_loc2', 'tem_in_loc3','hum_in_loc3', 'tem_coil_loc3','tem_out_loc1', 'hum_out_loc1','cond_loc1', 'cond_loc2','cond_loc3']
data_val.columns = ['mea_ddhr','plant','loc','tem_in','hum_in','tem_coil','tem_out','hum_out','x24h_tma','cond_loc24','x48h_tma','cond_loc48']

# 기본 데이터는 냅두고, 이걸 복사해서 쓰는 것이 효율적이다.
# 여기서 python이 기본적으로 reference로 shellow copy를 한다는 점을 인지하고 deep copy를 해야한다.
data1_plant1 = data_raw_plant1.copy(deep = True)
data1_plant2 = data_raw_plant2.copy(deep = True)

# copy by reference를 한다는 점을 이용하여, 빈번하게 수정할 데이터를 한번에 list로 갖고 있자.
data_cleaner = [data1_plant1, data1_plant2, data_val]

# 6. Simple Check



## 1. Check INFO

In [ ]:
#preview data
print("========= 1공장 TRAIN INFO =========\n")
data_raw_plant1.info()
print("\n========= 2공장 TRAIN INFO =========\n")
data_raw_plant2.info()
print("\n========= VALIDATION INFO =========\n")
data_val.info()

========= 1공장 TRAIN INFO =========

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58749 entries, 0 to 58748
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   mea_ddhr       58749 non-null  object 
 1   tem_in_loc1    57879 non-null  float64
 2   hum_in_loc1    57879 non-null  float64
 3   tem_coil_loc1  57879 non-null  float64
 4   tem_in_loc2    58629 non-null  float64
 5   hum_in_loc2    58629 non-null  float64
 6   tem_coil_loc2  58629 non-null  float64
 7   tem_in_loc3    58604 non-null  float64
 8   hum_in_loc3    58604 non-null  float64
 9   tem_coil_loc3  58629 non-null  float64
 10  tem_out_loc1   58629 non-null  float64
 11  hum_out_loc1   58629 non-null  float64
 12  cond_loc1      57879 non-null  float64
 13  cond_loc2      58629 non-null  float64
 14  cond_loc3      58604 non-null  float64
dtypes: float64(14), object(1)
memory usage: 6.7+ MB

========= 2공장 TRAIN INFO =========

<class 'pandas.co

In [ ]:
print("========= 1공장 TRAIN SAMPLE =========\n")
data_raw_plant1.sample(10).transpose() # random으로 10개를 뽑아준다.

========= 1공장 TRAIN SAMPLE =========



15535  ...             16654
mea_ddhr       2018-05-18 14:30  ...  2018-06-10 22:00
tem_in_loc1               21.26  ...             25.78
hum_in_loc1               67.95  ...             53.94
tem_coil_loc1             21.32  ...             24.29
tem_in_loc2               21.26  ...             24.59
hum_in_loc2               67.64  ...             56.38
tem_coil_loc2             21.58  ...             23.57
tem_in_loc3               20.53  ...             23.68
hum_in_loc3               71.31  ...             56.78
tem_coil_loc3             21.16  ...             23.11
tem_out_loc1              16.65  ...             22.06
hum_out_loc1              85.04  ...             57.94
cond_loc1                     0  ...                 0
cond_loc2                     0  ...                 0
cond_loc3                     0  ...                 0

[15 rows x 10 columns]

In [ ]:
print("========= 2공장 TRAIN SAMPLE =========\n")
data_raw_plant2.sample(10).transpose() # random으로 10개를 뽑아준다.

========= 2공장 TRAIN SAMPLE =========



55665  ...            40975
mea_ddhr       2019-03-10 6:50  ...  2018-11-27 6:50
tem_in_loc1               8.11  ...             8.72
hum_in_loc1               45.6  ...            60.84
tem_coil_loc1                8  ...             9.45
tem_in_loc2               9.08  ...            10.03
hum_in_loc2              43.22  ...            57.94
tem_coil_loc2             9.07  ...            10.75
tem_in_loc3                9.6  ...            10.15
hum_in_loc3              40.35  ...            57.23
tem_coil_loc3             10.6  ...            11.82
tem_out_loc1              4.96  ...             3.74
hum_out_loc1             60.35  ...            91.94
cond_loc1                    0  ...                0
cond_loc2                    0  ...                0
cond_loc3                    0  ...                0

[15 rows x 10 columns]

In [ ]:
print("========= VALIDATION SAMPLE =========\n")
data_val.sample(10).transpose() # random으로 10개를 뽑아준다.

========= VALIDATION SAMPLE =========



405   ...              3275
mea_ddhr    2019-05-13 23:00  ...  2020-03-02 21:00
plant                      2  ...                 1
loc                        1  ...                 3
tem_in                 18.15  ...              8.57
hum_in                 61.75  ...             43.83
tem_coil               18.95  ...              8.46
tem_out                15.98  ...              5.21
hum_out                74.02  ...             60.35
x24h_tma    2019-05-14 23:00  ...  2020-03-03 21:00
cond_loc24               NaN  ...               NaN
x48h_tma    2019-05-15 23:00  ...  2020-03-04 21:00
cond_loc48               NaN  ...               NaN

[12 rows x 10 columns]

## 2. Check NULL

In [ ]:
def countNull(df):
  """
  describe : 데이터 갯수, NULL 갯수, 백분율을 보기위한 코드
  <input>
  pd.dataframe
  <output>
  pd.dataframe
  """
  count_null = pd.DataFrame()
  count_null['numOfTotal'] = [len(df) for x in df.columns]
  count_null['numOfNull'] = list(df.isnull().sum())
  count_null['numOfNotnull'] = count_null['numOfTotal'] - count_null['numOfNull']
  count_null['percent'] = count_null['numOfNull']/count_null['numOfTotal']*100
  count_null['dataType'] = [type(df[x][0]) for x in df.columns]
  count_null.index = df.columns
  return count_null

In [ ]:
# plant1의 NULL 갯수
count_null_plant1 = countNull(data_raw_plant1)
count_null_plant1

numOfTotal  numOfNull  ...   percent                 dataType
mea_ddhr            58749          0  ...  0.000000            <class 'str'>
tem_in_loc1         58749        870  ...  1.480876  <class 'numpy.float64'>
hum_in_loc1         58749        870  ...  1.480876  <class 'numpy.float64'>
tem_coil_loc1       58749        870  ...  1.480876  <class 'numpy.float64'>
tem_in_loc2         58749        120  ...  0.204259  <class 'numpy.float64'>
hum_in_loc2         58749        120  ...  0.204259  <class 'numpy.float64'>
tem_coil_loc2       58749        120  ...  0.204259  <class 'numpy.float64'>
tem_in_loc3         58749        145  ...  0.246813  <class 'numpy.float64'>
hum_in_loc3         58749        145  ...  0.246813  <class 'numpy.float64'>
tem_coil_loc3       58749        120  ...  0.204259  <class 'numpy.float64'>
tem_out_loc1        58749        120  ...  0.204259  <class 'numpy.float64'>
hum_out_loc1        58749        120  ...  0.204259  <class 'numpy.float64'>
cond_loc1           58749        870  ...  1.480876  <class 'numpy.float64'>
cond_loc2           58749        120  ...  0.204259  <class 'numpy.float64'>
cond_loc3           58749        145  ...  0.246813  <class 'numpy.float64'>

[15 rows x 5 columns]

In [ ]:
# plant2의 NULL 갯수
count_null_plant2 = countNull(data_raw_plant2)
count_null_plant2

numOfTotal  numOfNull  ...   percent                 dataType
mea_ddhr            58792          0  ...  0.000000            <class 'str'>
tem_in_loc1         58792        374  ...  0.636141  <class 'numpy.float64'>
hum_in_loc1         58792        374  ...  0.636141  <class 'numpy.float64'>
tem_coil_loc1       58792        374  ...  0.636141  <class 'numpy.float64'>
tem_in_loc2         58792        374  ...  0.636141  <class 'numpy.float64'>
hum_in_loc2         58792        374  ...  0.636141  <class 'numpy.float64'>
tem_coil_loc2       58792        374  ...  0.636141  <class 'numpy.float64'>
tem_in_loc3         58792        374  ...  0.636141  <class 'numpy.float64'>
hum_in_loc3         58792        374  ...  0.636141  <class 'numpy.float64'>
tem_coil_loc3       58792        374  ...  0.636141  <class 'numpy.float64'>
tem_out_loc1        58792        374  ...  0.636141  <class 'numpy.float64'>
hum_out_loc1        58792        374  ...  0.636141  <class 'numpy.float64'>
cond_loc1           58792        374  ...  0.636141  <class 'numpy.float64'>
cond_loc2           58792        374  ...  0.636141  <class 'numpy.float64'>
cond_loc3           58792        374  ...  0.636141  <class 'numpy.float64'>

[15 rows x 5 columns]

plant2 데이터를 보면 모든 변수의 널의 갯수가 동일함을 알수 있다.

따라서 같은 row에 모든 널값이 있는지 확인한다.

In [ ]:
# 확인하는 코드
sum(data_raw_plant2.isnull().sum(axis=1) > 0)

374

결과적으로 null값은 모두 동일한 row에 분포 되어있음을 알수 있다.

즉 374개 정답값을 제외한 변수가 존재 하지 않음으로 제거하는것이 옳다.

## 3. Check Label Distribution

In [ ]:
def checkLabel(df, plant):
  label = pd.DataFrame()
  label['true'] = [sum(df['cond_loc1'] == 1),
                   sum(df['cond_loc2'] == 1),
                   sum(df['cond_loc3'] == 1)]
  label['false'] = [sum(df['cond_loc1'] == 0),
                   sum(df['cond_loc2'] == 0),
                   sum(df['cond_loc3'] == 0)]
  label['percent'] = label['true'] / (label['true'] + label['false']) * 100
  label.index = ['loc1', 'loc2', 'loc3']
  return label

In [ ]:
# plant1 loc에 대한 분포
label1 = checkLabel(data_raw_plant1, 1)
label1

true  false   percent
loc1   275  57604  0.475129
loc2   487  58142  0.830647
loc3   645  57959  1.100607

In [ ]:
# plant2 loc에 대한 분포
label2 = checkLabel(data_raw_plant2, 2)
label2

true  false   percent
loc1   277  58141  0.474169
loc2   277  58141  0.474169
loc3    44  58374  0.075319

# 7. Data Arrange

## 1. 사용하는 함수

In [ ]:
def arrangeData(df, plant):
  """
  describe : plant, loc에 따라 column을 볼 수 있게 데이터 프레임을 만든다.
  """
  ret_columns_name = ['mea_ddhr','plant','loc','tem_in','hum_in','tem_coil','tem_out_loc','hum_out_loc','cond_loc']
  ret_df = pd.DataFrame(columns=ret_columns_name)
  # plant = int(df.columns[1].split('_')[0][-1])
  np_df = np.array(df)
  ret_df = np.repeat(' '*30, len(df)*3*9).reshape(len(df)*3, 9)
  for index in range(len(df)):
    for loc in range(3):
      p = 3*loc + 1
      ret_df[3*index + loc,0] = np_df[index, 0]
      ret_df[3*index + loc,1] = plant
      ret_df[3*index + loc,2] = loc + 1
      ret_df[3*index + loc,3] = np_df[index, p] # tem_in
      ret_df[3*index + loc,4] = np_df[index, p+1] # hum_in
      ret_df[3*index + loc,5] = np_df[index, p+2] # tem_coil
      ret_df[3*index + loc,6] = np_df[index, 10] # tem_out
      ret_df[3*index + loc,7] = np_df[index, 11] # hum_out
      ret_df[3*index + loc,8] = np_df[index, 12+loc]
  ret_df = pd.DataFrame(ret_df, columns=ret_columns_name)
  for col in ret_columns_name:
    if col == 'mea_ddhr':
      ret_df[col] = ret_df[col].apply(lambda x : str(x) if x != 'nan' else np.nan)
    else:
      ret_df[col] = ret_df[col].apply(lambda x : np.float64(x) if x != 'nan' else np.nan)
  return ret_df

In [ ]:
import datetime
def stringToDatetime(df, colName='mea_ddhr'):
  """
  describe : string 을 datetime으로 변경한다.
  input = pd.Dataframe
  output = pd.Dataframe
  """
  if type(df[colName][0]) != str: # 이미 자료형이 datetime인 경우 return 한다.
    return df
  if len(df[colName][0].split(' ')[1].split(':')) == 2: # 날짜 정보에 초에 대한 정보가 없을 경우 초를 붙여주고 datetime으로 변경한다.
    df[colName] = df[colName].apply(lambda x : x + ':00' if x != 'nan' else np.nan)
  df[colName] = df[colName].apply(lambda x : datetime.datetime.strptime(x, '%Y-%m-%d %H:%M:%S') if x != 'nan' else np.nan)
  return df

def createDateTime(df, colName='mea_ddhr'):
  """
  describe : datetime의 정보에서 date 정보와 0시 기준 몇 분인지를 나타내는 time을 분리하여 추가한다.
  input = pd.Dataframe
  output = pd.Dataframe
  """
  df = stringToDatetime(df, colName)
  df['date'] = df[colName].apply(lambda x : x.date())
  df['time'] = df[colName].apply(lambda x : np.float64(x.hour * 60 + x.minute))
  return df

## 2. 데이터 형식 변경

In [ ]:
# 데이터 형식 변경
data_plant1 = arrangeData(data_raw_plant1, 1).copy(deep = True)
data_plant2 = arrangeData(data_raw_plant2, 2).copy(deep = True)

## 3. Date, time 추가

In [ ]:
data_plant1 = createDateTime(data_plant1)
data_plant2 = createDateTime(data_plant2)

In [ ]:
data_plant1

mea_ddhr  plant  loc  ...  cond_loc        date    time
0      2016-04-01 00:00:00    1.0  1.0  ...       0.0  2016-04-01     0.0
1      2016-04-01 00:00:00    1.0  2.0  ...       0.0  2016-04-01     0.0
2      2016-04-01 00:00:00    1.0  3.0  ...       0.0  2016-04-01     0.0
3      2016-04-01 03:00:00    1.0  1.0  ...       0.0  2016-04-01   180.0
4      2016-04-01 03:00:00    1.0  2.0  ...       0.0  2016-04-01   180.0
...                    ...    ...  ...  ...       ...         ...     ...
176242 2019-03-31 23:40:00    1.0  2.0  ...       0.0  2019-03-31  1420.0
176243 2019-03-31 23:40:00    1.0  3.0  ...       0.0  2019-03-31  1420.0
176244 2019-03-31 23:50:00    1.0  1.0  ...       0.0  2019-03-31  1430.0
176245 2019-03-31 23:50:00    1.0  2.0  ...       0.0  2019-03-31  1430.0
176246 2019-03-31 23:50:00    1.0  3.0  ...       0.0  2019-03-31  1430.0

[176247 rows x 11 columns]

In [ ]:
data_plant2

mea_ddhr  plant  loc  ...  cond_loc        date    time
0      2016-04-01 00:00:00    2.0  1.0  ...       0.0  2016-04-01     0.0
1      2016-04-01 00:00:00    2.0  2.0  ...       0.0  2016-04-01     0.0
2      2016-04-01 00:00:00    2.0  3.0  ...       0.0  2016-04-01     0.0
3      2016-04-01 03:00:00    2.0  1.0  ...       0.0  2016-04-01   180.0
4      2016-04-01 03:00:00    2.0  2.0  ...       0.0  2016-04-01   180.0
...                    ...    ...  ...  ...       ...         ...     ...
176371 2019-03-31 23:40:00    2.0  2.0  ...       0.0  2019-03-31  1420.0
176372 2019-03-31 23:40:00    2.0  3.0  ...       0.0  2019-03-31  1420.0
176373 2019-03-31 23:50:00    2.0  1.0  ...       NaN  2019-03-31  1430.0
176374 2019-03-31 23:50:00    2.0  2.0  ...       NaN  2019-03-31  1430.0
176375 2019-03-31 23:50:00    2.0  3.0  ...       NaN  2019-03-31  1430.0

[176376 rows x 11 columns]

## 4. TEST DATA 형식 변경

In [ ]:
tma24 = pd.DataFrame(data_val[['x24h_tma','plant','loc']])
tma24.columns = ['mea_ddhr','plant','loc']

tma48 = pd.DataFrame(data_val[['x48h_tma','plant','loc']])
tma48.columns = ['mea_ddhr','plant','loc']

data_test = pd.concat([tma24, tma48], ignore_index=True)
data_test = data_test.drop_duplicates()
data_test

mea_ddhr  plant  loc
0      2019-04-02 0:00      2    3
1      2019-04-02 4:30      1    2
2     2019-04-02 10:30      2    3
3     2019-04-02 11:00      1    1
4     2019-04-02 13:30      2    1
...                ...    ...  ...
7073   2020-03-31 3:00      1    2
7074   2020-03-31 5:00      2    3
7075   2020-03-31 7:00      2    1
7076  2020-03-31 11:30      1    1
7077  2020-03-31 19:30      1    1

[6753 rows x 3 columns]

In [ ]:
data_test = createDateTime(data_test)
data_test

mea_ddhr  plant  loc        date    time
0    2019-04-02 00:00:00      2    3  2019-04-02     0.0
1    2019-04-02 04:30:00      1    2  2019-04-02   270.0
2    2019-04-02 10:30:00      2    3  2019-04-02   630.0
3    2019-04-02 11:00:00      1    1  2019-04-02   660.0
4    2019-04-02 13:30:00      2    1  2019-04-02   810.0
...                  ...    ...  ...         ...     ...
7073 2020-03-31 03:00:00      1    2  2020-03-31   180.0
7074 2020-03-31 05:00:00      2    3  2020-03-31   300.0
7075 2020-03-31 07:00:00      2    1  2020-03-31   420.0
7076 2020-03-31 11:30:00      1    1  2020-03-31   690.0
7077 2020-03-31 19:30:00      1    1  2020-03-31  1170.0

[6753 rows x 5 columns]

# 7. 결측치 다루기

In [ ]:
is_NaN = data_plant1.isnull()
row_has_NaN = is_NaN.any(axis=1)
data_plant1[row_has_NaN]

mea_ddhr  plant  loc  ...  cond_loc        date    time
2304   2016-07-06 00:00:00    1.0  1.0  ...       NaN  2016-07-06     0.0
2305   2016-07-06 00:00:00    1.0  2.0  ...       NaN  2016-07-06     0.0
2306   2016-07-06 00:00:00    1.0  3.0  ...       NaN  2016-07-06     0.0
2931   2016-08-01 03:00:00    1.0  1.0  ...       NaN  2016-08-01   180.0
2932   2016-08-01 03:00:00    1.0  2.0  ...       NaN  2016-08-01   180.0
...                    ...    ...  ...  ...       ...         ...     ...
146005 2019-01-20 23:50:00    1.0  2.0  ...       NaN  2019-01-20  1430.0
146006 2019-01-20 23:50:00    1.0  3.0  ...       NaN  2019-01-20  1430.0
146436 2019-01-21 23:50:00    1.0  1.0  ...       NaN  2019-01-21  1430.0
146437 2019-01-21 23:50:00    1.0  2.0  ...       NaN  2019-01-21  1430.0
146438 2019-01-21 23:50:00    1.0  3.0  ...       NaN  2019-01-21  1430.0

[1135 rows x 11 columns]

## 1. Plant1에 대한 결측치 제거

먼저, 결측치가 함께 없는 row에 대해 제거를 수행한다.

### 1. Loc 1.0

In [ ]:
# nan의 개수가 이상함을 감지.
a = data_plant1[data_plant1["cond_loc"].isnull()]
b = data_plant1[data_plant1["tem_in"].isnull()]
c = data_plant1[data_plant1["hum_in"].isnull()]
d = data_plant1[data_plant1["tem_coil"].isnull()]
a = a[a["loc"] == 1.0]
b = b[b["loc"] == 1.0]
c = c[c["loc"] == 1.0]
d = d[d["loc"] == 1.0]
print(a.shape[0])
print(b.shape[0])
print(c.shape[0])
print(d.shape[0])


870
870
870
870


In [ ]:
# loc == 1.0일 때, tem_in, hum_in, tem_coil, cond_loc이 nan인 row는 같은 row이다. 검증하는 코드
a = data_plant1[["loc", "tem_in", "hum_in", "tem_coil", "cond_loc"]]
a = a[a["loc"] == 1.0]
a = a[["tem_in", "hum_in", "tem_coil", "cond_loc"]]
is_NaN = a.isnull()
row_has_NaN = is_NaN.all(axis=1)
print("870 * 4 = ", a[row_has_NaN].shape[0]*a[row_has_NaN].shape[1])



870 * 4 =  3480


### 2. Loc 2.0

In [ ]:
# nan의 개수가 이상함을 감지.
a = data_plant1[data_plant1["cond_loc"].isnull()]
b = data_plant1[data_plant1["tem_in"].isnull()]
c = data_plant1[data_plant1["hum_in"].isnull()]
d = data_plant1[data_plant1["tem_coil"].isnull()]
a = a[a["loc"] == 2.0]
b = b[b["loc"] == 2.0]
c = c[c["loc"] == 2.0]
d = d[d["loc"] == 2.0]
print(a.shape[0])
print(b.shape[0])
print(c.shape[0])
print(d.shape[0])


120
120
120
120


In [ ]:
# loc == 1.0일 때, tem_in, hum_in, tem_coil, cond_loc이 nan인 row는 같은 row이다. 검증하는 코드
a = data_plant1[["loc", "tem_in", "hum_in", "tem_coil", "cond_loc"]]
a = a[a["loc"] == 2.0]
a = a[["tem_in", "hum_in", "tem_coil", "cond_loc"]]
is_NaN = a.isnull()
row_has_NaN = is_NaN.all(axis=1)
print("120 * 4 = ", a[row_has_NaN].shape[0]*a[row_has_NaN].shape[1])



120 * 4 =  480


### 3. Loc 3.0

In [ ]:
# nan의 개수가 이상함을 감지.
a = data_plant1[data_plant1["cond_loc"].isnull()]
b = data_plant1[data_plant1["tem_in"].isnull()]
c = data_plant1[data_plant1["hum_in"].isnull()]
d = data_plant1[data_plant1["tem_coil"].isnull()]
a = a[a["loc"] == 3.0]
b = b[b["loc"] == 3.0]
c = c[c["loc"] == 3.0]
d = d[d["loc"] == 3.0]
print(a.shape[0])
print(b.shape[0])
print(c.shape[0])
print(d.shape[0])


145
145
145
120


tem_coil 가 nan인 개수가 보다 적다.

In [ ]:
# loc == 1.0일 때, tem_in, hum_in, tem_coil, cond_loc이 nan인 row는 같은 row이다. 검증하는 코드
a = data_plant1[["loc", "tem_in", "hum_in", "tem_coil", "cond_loc"]]
a = a[a["loc"] == 3.0]
a = a[["tem_in", "hum_in", "tem_coil", "cond_loc"]]
is_NaN = a.isnull()
row_has_NaN = is_NaN.all(axis=1)
print("120 * 4 = ", a[row_has_NaN].shape[0]*a[row_has_NaN].shape[1])



120 * 4 =  480


### 4. cond_loc(y) 값이 없는 경우 제거

In [ ]:
data_plant1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176247 entries, 0 to 176246
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   mea_ddhr     176247 non-null  datetime64[ns]
 1   plant        176247 non-null  float64       
 2   loc          176247 non-null  float64       
 3   tem_in       175112 non-null  float64       
 4   hum_in       175112 non-null  float64       
 5   tem_coil     175137 non-null  float64       
 6   tem_out_loc  175887 non-null  float64       
 7   hum_out_loc  175887 non-null  float64       
 8   cond_loc     175112 non-null  float64       
 9   date         176247 non-null  object        
 10  time         176247 non-null  float64       
dtypes: datetime64[ns](1), float64(9), object(1)
memory usage: 14.8+ MB


In [ ]:
# 그렇다면, cond_loc이 없는 경우에 대해 모두 삭제를 수행한다.
data_plant1 = data_plant1.dropna(subset = ["cond_loc"])
data_plant1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 175112 entries, 0 to 176246
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   mea_ddhr     175112 non-null  datetime64[ns]
 1   plant        175112 non-null  float64       
 2   loc          175112 non-null  float64       
 3   tem_in       175112 non-null  float64       
 4   hum_in       175112 non-null  float64       
 5   tem_coil     175112 non-null  float64       
 6   tem_out_loc  175112 non-null  float64       
 7   hum_out_loc  175112 non-null  float64       
 8   cond_loc     175112 non-null  float64       
 9   date         175112 non-null  object        
 10  time         175112 non-null  float64       
dtypes: datetime64[ns](1), float64(9), object(1)
memory usage: 16.0+ MB


## 2. Plant2에 대한 결측치 제거

In [ ]:
# 그렇다면, cond_loc이 없는 경우에 대해 모두 삭제를 수행한다.
data_plant2 = data_plant2.dropna(subset = ["cond_loc"])
data_plant2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 175254 entries, 0 to 176372
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   mea_ddhr     175254 non-null  datetime64[ns]
 1   plant        175254 non-null  float64       
 2   loc          175254 non-null  float64       
 3   tem_in       175254 non-null  float64       
 4   hum_in       175254 non-null  float64       
 5   tem_coil     175254 non-null  float64       
 6   tem_out_loc  175254 non-null  float64       
 7   hum_out_loc  175254 non-null  float64       
 8   cond_loc     175254 non-null  float64       
 9   date         175254 non-null  object        
 10  time         175254 non-null  float64       
dtypes: datetime64[ns](1), float64(9), object(1)
memory usage: 16.0+ MB


y값이 없는 경우를 다 지우고 나니, 결측치가 삭제되었다.

# 8. 데이터 저장하기


In [ ]:
# 정제된 데이터를 저장한다.
data_plant1.to_csv(os.path.join(DATA_CONV_PATH, 'plant1_by_loc_with_datetime.csv'))
data_plant2.to_csv(os.path.join(DATA_CONV_PATH, 'plant2_by_loc_with_datetime.csv'))
data_test.to_csv(os.path.join(DATA_CONV_PATH, 'plant_test_with_datetime.csv'))